File: jejecrunch/programmers/17680.java
import java.util.*;

class Solution {
    static final int CACHE_HIT = 1;
    static final int CACHE_MISS = 5;
    static ArrayList<String> memory = new ArrayList<>();

    public int solution(int cacheSize, String[] cities) {
        int answer = 0;
        if (cacheSize == 0)
            return cities.length * CACHE_MISS;

        for (int i = 0; i < cities.length; i++) {
            String temp = cities[i].toUpperCase(); // 대소문자 구분 X
            if (memory.remove(temp)) {
                answer += CACHE_HIT;
                memory.add(temp);
            } else {
                answer += CACHE_MISS;
                if (memory.size() >= cacheSize) {
                    memory.remove(0);
                }
                memory.add(temp);
            }
        }

        return answer;
    }

}

File: jejecrunch/programmers/javascript/17680.js
function solution(cacheSize, cities) {
    var answer = 0
    const CACHE_HIT = 1,
        CACHE_MISS = 5
    let memory = []

    if (cacheSize == 0) return cities.length * CACHE_MISS

    for (let i = 0; i < cities.length; i++) {
        let temp = cities[i].toUpperCase()
        if (memory.findIndex((v) => v == temp) > -1) {
            memory = memory.filter((v) => v != temp)
            answer += CACHE_HIT
            memory.push(temp)
        } else {
            answer += CACHE_MISS
            if (memory.length >= cacheSize) {
                memory.shift()
            }
            memory.push(temp)
        }
    }
    return answer
}
